File: packages/openducktor-mcp/src/task-resolution.ts
import { sanitizeSlug } from "./beads-runtime";
import type { TaskCard } from "./contracts";

export interface TaskIndex {
  tasks: TaskCard[];
  entries: TaskIndexEntry[];
  idExact: Map<string, TaskCard>;
  idLower: Map<string, TaskCard[]>;
  idSuffix: Map<string, TaskCard[]>;
  titleExact: Map<string, TaskCard[]>;
  titleSlug: Map<string, TaskCard[]>;
}

interface TaskIndexEntry {
  task: TaskCard;
  idLower: string;
  titleLower: string;
  titleSlug: string;
}

const MAX_TASK_CANDIDATES = 5;

const formatTaskRef = (task: TaskCard): string => `${task.id} (${task.title})`;

export class TaskResolutionAmbiguousError extends Error {
  readonly requestedTaskId: string;
  readonly candidates: string[];

  constructor(requestedTaskId: string, candidates: string[]) {
    super(
      `Task identifier "${requestedTaskId}" is ambiguous. Use exact task id. Candidates: ${candidates.join(", ")}`,
    );
    this.name = "TaskResolutionAmbiguousError";
    this.requestedTaskId = requestedTaskId;
    this.candidates = candidates;
  }
}

export class TaskResolutionNotFoundError extends Error {
  readonly requestedTaskId: string;
  readonly candidates: string[];

  constructor(requestedTaskId: string, candidates: string[]) {
    const hintSuffix = candidates.length > 0 ? ` Candidate task ids: ${candidates.join(", ")}` : "";
    super(`Task not found: ${requestedTaskId}.${hintSuffix}`);
    this.name = "TaskResolutionNotFoundError";
    this.requestedTaskId = requestedTaskId;
    this.candidates = candidates;
  }
}

export const normalizeTitleKey = (value: string): string => value.trim().toLowerCase();

export const toSearchSlug = (value: string): string => {
  if (!/[a-z0-9]/i.test(value)) {
    return "";
  }
  return sanitizeSlug(value);
};

function throwAmbiguousTaskIdentifier(requestedTaskId: string, matches: TaskCard[]): never {
  const candidates = matches.slice(0, MAX_TASK_CANDIDATES).map(formatTaskRef);
  throw new TaskResolutionAmbiguousError(requestedTaskId, candidates);
}

export function buildTaskIndex(tasks: TaskCard[]): TaskIndex {
  const idExact = new Map<string, TaskCard>();
  const idLower = new Map<string, TaskCard[]>();
  const idSuffix = new Map<string, TaskCard[]>();
  const titleExact = new Map<string, TaskCard[]>();
  const titleSlug = new Map<string, TaskCard[]>();
  const entries: TaskIndexEntry[] = [];

  const addTaskToBucket = (map: Map<string, TaskCard[]>, key: string, task: TaskCard): void => {
    const existing = map.get(key);
    if (existing) {
      existing.push(task);
    } else {
      map.set(key, [task]);
    }
  };

  for (const task of tasks) {
    const normalizedId = normalizeTitleKey(task.id);
    const normalizedTitle = normalizeTitleKey(task.title);
    const normalizedTitleSlug = toSearchSlug(task.title);

    idExact.set(task.id, task);
    addTaskToBucket(idLower, normalizedId, task);

    addTaskToBucket(idSuffix, normalizedId, task);
    for (let i = 0; i < normalizedId.length; i += 1) {
      if (normalizedId[i] !== "-") {
        continue;
      }
      const suffix = normalizedId.slice(i + 1);
      if (suffix.length > 0) {
        addTaskToBucket(idSuffix, suffix, task);
      }
    }

    addTaskToBucket(titleExact, normalizedTitle, task);
    if (normalizedTitleSlug.length > 0) {
      addTaskToBucket(titleSlug, normalizedTitleSlug, task);
    }

    entries.push({
      task,
      idLower: normalizedId,
      titleLower: normalizedTitle,
      titleSlug: normalizedTitleSlug,
    });
  }

  return {
    tasks,
    entries,
    idExact,
    idLower,
    idSuffix,
    titleExact,
    titleSlug,
  };
}

export function resolveTaskFromIndex(index: TaskIndex, requestedTaskId: string): TaskCard {
  const requestedLiteral = requestedTaskId.trim();
  if (requestedLiteral.length === 0) {
    throw new Error("Missing taskId.");
  }

  const requestedLower = normalizeTitleKey(requestedLiteral);
  const requestedSlug = toSearchSlug(requestedLiteral);

  const exact = index.idExact.get(requestedLiteral);
  if (exact) {
    return exact;
  }

  const byCaseInsensitiveId = index.idLower.get(requestedLower);
  if (byCaseInsensitiveId) {
    if (byCaseInsensitiveId.length === 1 && byCaseInsensitiveId[0]) {
      return byCaseInsensitiveId[0];
    }
    if (byCaseInsensitiveId.length > 1) {
      throwAmbiguousTaskIdentifier(requestedTaskId, byCaseInsensitiveId);
    }
  }

  if (requestedSlug.length > 0) {
    const byIdSuffix = index.idSuffix.get(requestedSlug);
    if (byIdSuffix?.length === 1 && byIdSuffix[0]) {
      return byIdSuffix[0];
    }
    if (byIdSuffix && byIdSuffix.length > 1) {
      throwAmbiguousTaskIdentifier(requestedTaskId, byIdSuffix);
    }
  }

  const byTitleExact = index.titleExact.get(requestedLower);
  if (byTitleExact) {
    if (byTitleExact.length === 1 && byTitleExact[0]) {
      return byTitleExact[0];
    }
    if (byTitleExact.length > 1) {
      throwAmbiguousTaskIdentifier(requestedTaskId, byTitleExact);
    }
  }

  if (requestedSlug.length > 0) {
    const byTitleSlugExact = index.titleSlug.get(requestedSlug);
    if (byTitleSlugExact) {
      if (byTitleSlugExact.length === 1 && byTitleSlugExact[0]) {
        return byTitleSlugExact[0];
      }
      if (byTitleSlugExact.length > 1) {
        throwAmbiguousTaskIdentifier(requestedTaskId, byTitleSlugExact);
      }
    }
  }

  const byTitleContains: TaskCard[] = [];
  if (requestedLower.length > 0 || requestedSlug.length > 0) {
    for (const entry of index.entries) {
      const matchesLower = requestedLower.length > 0 && entry.titleLower.includes(requestedLower);
      const matchesSlug = requestedSlug.length > 0 && entry.titleSlug.includes(requestedSlug);

      if (matchesLower || matchesSlug) {
        byTitleContains.push(entry.task);
        if (byTitleContains.length > MAX_TASK_CANDIDATES) {
          break;
        }
      }
    }
  }

  if (byTitleContains.length === 1 && byTitleContains[0]) {
    return byTitleContains[0];
  }
  if (byTitleContains.length > 1) {
    throwAmbiguousTaskIdentifier(requestedTaskId, byTitleContains);
  }

  const hints: TaskCard[] = [];
  if (requestedLower.length > 0 || requestedSlug.length > 0) {
    for (const entry of index.entries) {
      const matchesIdLower = requestedLower.length > 0 && entry.idLower.includes(requestedLower);
      const matchesTitleLower =
        requestedLower.length > 0 && entry.titleLower.includes(requestedLower);
      const matchesIdSlug = requestedSlug.length > 0 && entry.idLower.includes(requestedSlug);
      const matchesTitleSlug = requestedSlug.length > 0 && entry.titleSlug.includes(requestedSlug);

      if (matchesIdLower || matchesTitleLower || matchesIdSlug || matchesTitleSlug) {
        hints.push(entry.task);
        if (hints.length >= MAX_TASK_CANDIDATES) {
          break;
        }
      }
    }
  }

  const fallback = (hints.length > 0 ? hints : index.tasks.slice(0, MAX_TASK_CANDIDATES)).map(
    formatTaskRef,
  );
  throw new TaskResolutionNotFoundError(requestedTaskId, fallback);
}

File: apps/desktop/src-tauri/crates/host-application/src/app_service/odt_mcp/task_resolution.rs
use anyhow::{anyhow, Result};
use host_domain::TaskCard;

use crate::app_service::workflow_rules::normalize_title_key;

const MAX_TASK_CANDIDATES: usize = 5;

fn sanitize_slug(input: &str) -> String {
    let mut slug = String::new();
    let mut last_dash = false;

    for ch in input.chars() {
        let lower = ch.to_ascii_lowercase();
        if lower.is_ascii_alphanumeric() {
            slug.push(lower);
            last_dash = false;
            continue;
        }
        if !last_dash {
            slug.push('-');
            last_dash = true;
        }
    }

    slug.trim_matches('-').to_string()
}

fn format_task_ref(task: &TaskCard) -> String {
    format!("{} ({})", task.id, task.title)
}

fn throw_ambiguous_task_identifier(
    requested_task_id: &str,
    matches: &[TaskCard],
) -> Result<TaskCard> {
    let candidates = matches
        .iter()
        .take(MAX_TASK_CANDIDATES)
        .map(format_task_ref)
        .collect::<Vec<_>>()
        .join(", ");
    Err(anyhow!(
        "Task identifier \"{}\" is ambiguous. Use exact task id. Candidates: {}",
        requested_task_id,
        candidates
    ))
}

pub(super) fn resolve_task_reference(
    tasks: &[TaskCard],
    requested_task_id: &str,
) -> Result<TaskCard> {
    let requested_literal = requested_task_id.trim();
    if requested_literal.is_empty() {
        return Err(anyhow!("Missing taskId."));
    }

    let requested_lower = normalize_title_key(requested_literal);
    let requested_slug = sanitize_slug(requested_literal);

    if let Some(task) = tasks.iter().find(|task| task.id == requested_literal) {
        return Ok(task.clone());
    }

    let by_case_insensitive_id = tasks
        .iter()
        .filter(|task| normalize_title_key(&task.id) == requested_lower)
        .cloned()
        .collect::<Vec<_>>();
    if by_case_insensitive_id.len() == 1 {
        return Ok(by_case_insensitive_id[0].clone());
    }
    if by_case_insensitive_id.len() > 1 {
        return throw_ambiguous_task_identifier(requested_task_id, &by_case_insensitive_id);
    }

    if !requested_slug.is_empty() {
        // ID matching intentionally accepts any dash-separated segment so users can paste
        // short memorable fragments from generated task ids, not only trailing suffixes.
        let by_id_segment = tasks
            .iter()
            .filter(|task| {
                let normalized_id = normalize_title_key(&task.id);
                normalized_id == requested_slug
                    || normalized_id
                        .split('-')
                        .any(|suffix| !suffix.is_empty() && suffix == requested_slug)
            })
            .cloned()
            .collect::<Vec<_>>();
        if by_id_segment.len() == 1 {
            return Ok(by_id_segment[0].clone());
        }
        if by_id_segment.len() > 1 {
            return throw_ambiguous_task_identifier(requested_task_id, &by_id_segment);
        }
    }

    let by_title_exact = tasks
        .iter()
        .filter(|task| normalize_title_key(&task.title) == requested_lower)
        .cloned()
        .collect::<Vec<_>>();
    if by_title_exact.len() == 1 {
        return Ok(by_title_exact[0].clone());
    }
    if by_title_exact.len() > 1 {
        return throw_ambiguous_task_identifier(requested_task_id, &by_title_exact);
    }

    if !requested_slug.is_empty() {
        let by_title_slug_exact = tasks
            .iter()
            .filter(|task| sanitize_slug(&task.title) == requested_slug)
            .cloned()
            .collect::<Vec<_>>();
        if by_title_slug_exact.len() == 1 {
            return Ok(by_title_slug_exact[0].clone());
        }
        if by_title_slug_exact.len() > 1 {
            return throw_ambiguous_task_identifier(requested_task_id, &by_title_slug_exact);
        }
    }

    let by_title_contains = tasks
        .iter()
        .filter(|task| {
            let title_lower = normalize_title_key(&task.title);
            let title_slug = sanitize_slug(&task.title);
            (!requested_lower.is_empty() && title_lower.contains(&requested_lower))
                || (!requested_slug.is_empty() && title_slug.contains(&requested_slug))
        })
        .take(MAX_TASK_CANDIDATES + 1)
        .cloned()
        .collect::<Vec<_>>();
    if by_title_contains.len() == 1 {
        return Ok(by_title_contains[0].clone());
    }
    if by_title_contains.len() > 1 {
        return throw_ambiguous_task_identifier(requested_task_id, &by_title_contains);
    }

    let hints = tasks
        .iter()
        .filter(|task| {
            let id_lower = normalize_title_key(&task.id);
            let title_lower = normalize_title_key(&task.title);
            let title_slug = sanitize_slug(&task.title);
            (!requested_lower.is_empty()
                && (id_lower.contains(&requested_lower) || title_lower.contains(&requested_lower)))
                || (!requested_slug.is_empty()
                    && (id_lower.contains(&requested_slug) || title_slug.contains(&requested_slug)))
        })
        .take(MAX_TASK_CANDIDATES)
        .map(format_task_ref)
        .collect::<Vec<_>>();
    let candidates = if hints.is_empty() {
        tasks
            .iter()
            .take(MAX_TASK_CANDIDATES)
            .map(format_task_ref)
            .collect::<Vec<_>>()
    } else {
        hints
    };

    if candidates.is_empty() {
        return Err(anyhow!("Task not found: {}.", requested_task_id));
    }

    Err(anyhow!(
        "Task not found: {}. Candidate task ids: {}",
        requested_task_id,
        candidates.join(", ")
    ))
}
